## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [34]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [35]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [36]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [ ]:
df.sort_values(by=['client_id'])
clean_headers_df = df.rename(columns={'first':'First Name',
                   'last':'Last Name',
                   'job':'Occupation',
                   'phone':'Phone Number',
                   'email':'Email',
                   'client_id':'Client ID',
                   'order_id':'Order Number',
                   'order_date':'Order Date',
                   'order_week':'Order Week',
                   'order_year':'Order Year',
                   'item_id':'Item Number',
                   'category':'Category',
                   'subcategory':'Subcategory',
                   'unit_price':'Price',
                   'unit_cost':'Cost',
                   'unit_weight':'Weight',
                   'qty':'Quantity',
                   'line_number':'Line Number'})



First Name Last Name             Occupation        Phone Number  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 Email  Client ID  Order Number  Order Date  \
0     harding.donald.7185@sullivan.com      58515       8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609       8069089  2023-05-19   
2                swatson8146@payne.net      57113       1902144  2023-01-29   
3               bakernathan@benson.com      46554       9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089       1322274  2023-05-28   

   Order Week  Order Year     Item Number     Category         Subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

     Price    Cost  Weight  Quantity  Line Number  
0  1096.80  762.71    7.50       105            1  
1    24.95   15.09    1.49        21            0  
2    13.52    7.86    1.68        39            6  
3    36.42   24.85    1.23        29            3  
4   195.10  108.17   46.43        20            1

In [ ]:
clean_headers_df.sort_values('Client ID').head(20)
column_rearrange_df = clean_headers_df[['Client ID',
                  'First Name',
                  'Last Name',
                  'Phone Number',
                  'Email',
                  'Order Number',
                  'Order Date',
                  'Order Week',
                  'Order Year',
                  'Category',
                  'Subcategory',
                  'Item Number',
                  'Price',
                  'Cost',
                  'Quantity',
                  'Weight',
                  'Line Number']]


In [154]:
cust_corders_df = column_rearrange_df.sort_values(by=['Client ID','Order Date'])
cust_corders_df.head(20)

Client ID First Name Last Name  Phone Number  \
317        10033      David     Ochoa  281.867.1615   
20791      10033      David     Ochoa  281.867.1615   
38158      10033      David     Ochoa  281.867.1615   
49037      10033      David     Ochoa  281.867.1615   
41773      10033      David     Ochoa  281.867.1615   
42047      10033      David     Ochoa  281.867.1615   
44345      10033      David     Ochoa  281.867.1615   
47797      10033      David     Ochoa  281.867.1615   
48120      10033      David     Ochoa  281.867.1615   
4461       10033      David     Ochoa  281.867.1615   
19563      10033      David     Ochoa  281.867.1615   
35217      10033      David     Ochoa  281.867.1615   
38447      10033      David     Ochoa  281.867.1615   
44111      10033      David     Ochoa  281.867.1615   
44787      10033      David     Ochoa  281.867.1615   
45405      10033      David     Ochoa  281.867.1615   
46890      10033      David     Ochoa  281.867.1615   
53449      10033      David     Ochoa  281.867.1615   
1294       10033      David     Ochoa  281.867.1615   
9920       10091    Shannon   Jimenez    8523477742   

                                      Email  Order Number  Order Date  \
317    david.measure.3427@morales-young.org       8159112  2023-01-16   
20791  david.measure.3427@morales-young.org       8159112  2023-01-16   
38158  david.measure.3427@morales-young.org       8159112  2023-01-16   
49037  david.measure.3427@morales-young.org       8159112  2023-01-16   
41773  david.measure.3427@morales-young.org       3574869  2023-01-26   
42047  david.measure.3427@morales-young.org       3574869  2023-01-26   
44345  david.measure.3427@morales-young.org       3574869  2023-01-26   
47797  david.measure.3427@morales-young.org       3574869  2023-01-26   
48120  david.measure.3427@morales-young.org       3229787  2023-01-30   
4461   david.measure.3427@morales-young.org       9244566  2023-03-11   
19563  david.measure.3427@morales-young.org       9244566  2023-03-11   
35217  david.measure.3427@morales-young.org       9244566  2023-03-11   
38447  david.measure.3427@morales-young.org       9244566  2023-03-11   
44111  david.measure.3427@morales-young.org       9244566  2023-03-11   
44787  david.measure.3427@morales-young.org       9244566  2023-03-11   
45405  david.measure.3427@morales-young.org       9244566  2023-03-11   
46890  david.measure.3427@morales-young.org       9244566  2023-03-11   
53449  david.measure.3427@morales-young.org       9244566  2023-03-11   
1294   david.measure.3427@morales-young.org       1381771  2023-05-15   
9920             jimenez.shannon@palmer.com       8834547  2023-01-04   

       Order Week  Order Year     Category         Subcategory  \
317             3        2023    furniture              tables   
20791           3        2023  consumables      paper products   
38158           3        2023    furniture              stands   
49037           3        2023        decor            wall art   
41773           4        2023     software          accounting   
42047           4        2023        decor                misc   
44345           4        2023    furniture              stands   
47797           4        2023     software  project management   
48120           5        2023  electronics        audio visual   
4461           10        2023  consumables      paper products   
19563          10        2023  consumables   bathroom supplies   
35217          10        2023  consumables    kitchen supplies   
38447          10        2023  consumables    kitchen supplies   
44111          10        2023     software          accounting   
44787          10        2023    furniture               desks   
45405          10        2023  consumables    kitchen supplies   
46890          10        2023     software           logistics   
53449          10        2023    furniture              stands   
1294           20        2023  electronics             telecom 

In [155]:
# Use this space to do any additional research
# and familiarize yourself with the data.
cust_corders_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 54639 entries, 317 to 28537
Data columns (total 17 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Client ID     54639 non-null  int64  
 1   First Name    54639 non-null  object 
 2   Last Name     54639 non-null  object 
 3   Phone Number  54639 non-null  object 
 4   Email         54639 non-null  object 
 5   Order Number  54639 non-null  int64  
 6   Order Date    54639 non-null  object 
 7   Order Week    54639 non-null  int64  
 8   Order Year    54639 non-null  int64  
 9   Category      54639 non-null  object 
 10  Subcategory   54639 non-null  object 
 11  Item Number   54639 non-null  object 
 12  Price         54639 non-null  float64
 13  Cost          54639 non-null  float64
 14  Quantity      54639 non-null  int64  
 15  Weight        54639 non-null  float64
 16  Line Number   54639 non-null  int64  
dtypes: float64(3), int64(6), object(8)
memory usage: 7.5+ MB


In [156]:
# What three item categories had the most entries?
top_three_categories = cust_corders_df['Category'].value_counts().head(3)
print(top_three_categories)


Category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [157]:
# For the category with the most entries, which subcategory had the most entries?
top_category_count = top_three_categories.iloc[0]
#top_category = sorted_df['Category'].value_counts().idxmax()
print(f"Count of Consumables: {top_category_count}")
#consumables_count = sorted_df[sorted_df['Category'] == top_category_count]
#top_subcategory = consumables_list['Subcategory'].value_counts().idxmax()
print(f"consumables count: {top_category_count}")
print(f"top subcategory: {top_subcategory}")

Count of Consumables: 23538
consumables count: 23538
top subcategory: bathroom supplies


In [158]:
# Which five clients had the most entries in the data?
cust_corders_df['Client ID'].value_counts().head(5)

Client ID
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [159]:
# Store the client ids of those top 5 clients in a list.
top_clients_list = cust_corders_df['Client ID'].value_counts().head(5)
print(top_clients_list)             #why does this print the quantities(?) with no header?



Client ID
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [160]:
# How many total units (the qty column) did the client with the most entries order order?
top_clients_list = cust_corders_df['Client ID'].value_counts().head(5)
print(top_clients_list)   

Client ID
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [178]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
subtotal = (cust_corders_df['Price'] * cust_corders_df['Quantity']).round(2)
cust_corders_df['Subtotal'] = subtotal
cust_corders_df

Client ID First Name Last Name        Phone Number  \
317        10033      David     Ochoa        281.867.1615   
20791      10033      David     Ochoa        281.867.1615   
38158      10033      David     Ochoa        281.867.1615   
49037      10033      David     Ochoa        281.867.1615   
41773      10033      David     Ochoa        281.867.1615   
...          ...        ...       ...                 ...   
34807      99984   Samantha    Murray  534.390.2260x56734   
41584      99984   Samantha    Murray  534.390.2260x56734   
42531      99984   Samantha    Murray  534.390.2260x56734   
43694      99984   Samantha    Murray  534.390.2260x56734   
28537      99984   Samantha    Murray  534.390.2260x56734   

                                         Email  Order Number  Order Date  \
317       david.measure.3427@morales-young.org       8159112  2023-01-16   
20791     david.measure.3427@morales-young.org       8159112  2023-01-16   
38158     david.measure.3427@morales-young.org       8159112  2023-01-16   
49037     david.measure.3427@morales-young.org       8159112  2023-01-16   
41773     david.measure.3427@morales-young.org       3574869  2023-01-26   
...                                        ...           ...         ...   
34807  murray.samantha.4079@gonzalez-brown.com       5491843  2023-04-28   
41584  murray.samantha.4079@gonzalez-brown.com       5491843  2023-04-28   
42531  murray.samantha.4079@gonzalez-brown.com       5491843  2023-04-28   
43694  murray.samantha.4079@gonzalez-brown.com       5491843  2023-04-28   
28537  murray.samantha.4079@gonzalez-brown.com       5546127  2023-05-22   

       Order Week  Order Year     Category  ...   Price    Cost  Quantity  \
317             3        2023    furniture  ...  196.08  145.92      1100   
20791           3        2023  consumables  ...   91.05   77.83       327   
38158           3        2023    furniture  ...    1.48    1.22       146   
49037           3        2023        decor  ...  112.34   61.72        10   
41773           4        2023     software  ...  311.98  178.12         6   
...           ...         ...          ...  ...     ...     ...       ...   
34807          17        2023  electronics  ...  136.99  105.43       117   
41584          17        2023     software  ...   36.49   21.29        35   
42531          17        2023    furniture  ...    7.98    6.68        52   
43694          17        2023  consumables  ...   22.38   11.85        64   
28537          21        2023        decor  ...  485.97  294.19        58   

       Weight  Line Number   Subtotal  Shipping Price   Total Price  \
317      2.76            1  215688.00            27.6  215690.55300   
20791    0.83            0   29773.35             8.3   29774.11775   
38158   18.04            3     216.08           180.4     232.76700   
49037    8.50            2    1123.40            85.0    1131.26250   
41773    2.34            3    1871.88            23.4    1874.04450   
...       ...          ...        ...             ...           ...   
34807    2.47            0   16027.83            24.7   16030.11475   
41584   13.78            2    1277.15           137.8    1289.89650   
42531    7.15            1     414.96            71.5     421.57375   
43694    4.28            4    1432.32            42.8    1436.27900   
28537    2.93            0   28186.26            29.3   28188.97025   

       Line Cost  Item Cost  
317    160539.60  160539.60  
20791   25458.71   25458.71  
38158     358.52     358.52  
49037     702.20     702.20  
41773    1092.12    1092.12  
...          ...        ...  
34807   12360.01   12360.01  
41584     882.95     882.95  
42531     418.86     418.86  
43694     801.20     801.20  
28537   17092.32   17092.32  

[54639 rows x 22 columns]

In [179]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
cust_corders_df["Shipping Price"] = cust_corders_df["Weight"].apply(lambda x: x * 7 if x > 50 else x * 10)
cust_corders_df.head()



Client ID First Name Last Name  Phone Number  \
317        10033      David     Ochoa  281.867.1615   
20791      10033      David     Ochoa  281.867.1615   
38158      10033      David     Ochoa  281.867.1615   
49037      10033      David     Ochoa  281.867.1615   
41773      10033      David     Ochoa  281.867.1615   

                                      Email  Order Number  Order Date  \
317    david.measure.3427@morales-young.org       8159112  2023-01-16   
20791  david.measure.3427@morales-young.org       8159112  2023-01-16   
38158  david.measure.3427@morales-young.org       8159112  2023-01-16   
49037  david.measure.3427@morales-young.org       8159112  2023-01-16   
41773  david.measure.3427@morales-young.org       3574869  2023-01-26   

       Order Week  Order Year     Category  ...   Price    Cost  Quantity  \
317             3        2023    furniture  ...  196.08  145.92      1100   
20791           3        2023  consumables  ...   91.05   77.83       327   
38158           3        2023    furniture  ...    1.48    1.22       146   
49037           3        2023        decor  ...  112.34   61.72        10   
41773           4        2023     software  ...  311.98  178.12         6   

       Weight  Line Number   Subtotal  Shipping Price   Total Price  \
317      2.76            1  215688.00            27.6  215690.55300   
20791    0.83            0   29773.35             8.3   29774.11775   
38158   18.04            3     216.08           180.4     232.76700   
49037    8.50            2    1123.40            85.0    1131.26250   
41773    2.34            3    1871.88            23.4    1874.04450   

       Line Cost  Item Cost  
317    160539.60  160539.60  
20791   25458.71   25458.71  
38158     358.52     358.52  
49037     702.20     702.20  
41773    1092.12    1092.12  

[5 rows x 22 columns]

In [176]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
cust_corders_df['Total Price'] = cust_corders_df['Subtotal'] + cust_corders_df['Shipping Price'] * 0.0925
cust_corders_df

Client ID First Name Last Name        Phone Number  \
317        10033      David     Ochoa        281.867.1615   
20791      10033      David     Ochoa        281.867.1615   
38158      10033      David     Ochoa        281.867.1615   
49037      10033      David     Ochoa        281.867.1615   
41773      10033      David     Ochoa        281.867.1615   
...          ...        ...       ...                 ...   
34807      99984   Samantha    Murray  534.390.2260x56734   
41584      99984   Samantha    Murray  534.390.2260x56734   
42531      99984   Samantha    Murray  534.390.2260x56734   
43694      99984   Samantha    Murray  534.390.2260x56734   
28537      99984   Samantha    Murray  534.390.2260x56734   

                                         Email  Order Number  Order Date  \
317       david.measure.3427@morales-young.org       8159112  2023-01-16   
20791     david.measure.3427@morales-young.org       8159112  2023-01-16   
38158     david.measure.3427@morales-young.org       8159112  2023-01-16   
49037     david.measure.3427@morales-young.org       8159112  2023-01-16   
41773     david.measure.3427@morales-young.org       3574869  2023-01-26   
...                                        ...           ...         ...   
34807  murray.samantha.4079@gonzalez-brown.com       5491843  2023-04-28   
41584  murray.samantha.4079@gonzalez-brown.com       5491843  2023-04-28   
42531  murray.samantha.4079@gonzalez-brown.com       5491843  2023-04-28   
43694  murray.samantha.4079@gonzalez-brown.com       5491843  2023-04-28   
28537  murray.samantha.4079@gonzalez-brown.com       5546127  2023-05-22   

       Order Week  Order Year     Category  ...   Price    Cost  Quantity  \
317             3        2023    furniture  ...  196.08  145.92      1100   
20791           3        2023  consumables  ...   91.05   77.83       327   
38158           3        2023    furniture  ...    1.48    1.22       146   
49037           3        2023        decor  ...  112.34   61.72        10   
41773           4        2023     software  ...  311.98  178.12         6   
...           ...         ...          ...  ...     ...     ...       ...   
34807          17        2023  electronics  ...  136.99  105.43       117   
41584          17        2023     software  ...   36.49   21.29        35   
42531          17        2023    furniture  ...    7.98    6.68        52   
43694          17        2023  consumables  ...   22.38   11.85        64   
28537          21        2023        decor  ...  485.97  294.19        58   

       Weight  Line Number   Subtotal  Shipping Price   Total Price  \
317      2.76            1  215688.00            27.6  215690.55300   
20791    0.83            0   29773.35             8.3   29774.11775   
38158   18.04            3     216.08           180.4     232.76700   
49037    8.50            2    1123.40            85.0    1131.26250   
41773    2.34            3    1871.88            23.4    1874.04450   
...       ...          ...        ...             ...           ...   
34807    2.47            0   16027.83            24.7   16030.11475   
41584   13.78            2    1277.15           137.8    1289.89650   
42531    7.15            1     414.96            71.5     421.57375   
43694    4.28            4    1432.32            42.8    1436.27900   
28537    2.93            0   28186.26            29.3   28188.97025   

       Line Cost  Item Cost  
317    160539.60  160539.60  
20791   25458.71   25458.71  
38158     358.52     358.52  
49037     702.20     702.20  
41773    1092.12    1092.12  
...          ...        ...  
34807   12360.01   12360.01  
41584     882.95     882.95  
42531     418.86     418.86  
43694     801.20     801.20  
28537   17092.32   17092.32  

[54639 rows x 22 columns]

In [175]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

cust_corders_df['Item Cost'] = cust_corders_df['Cost'] * cust_corders_df['Quantity'] + cust_corders_df['Shipping Price']
cust_corders_df.head()

Client ID First Name Last Name  Phone Number  \
317        10033      David     Ochoa  281.867.1615   
20791      10033      David     Ochoa  281.867.1615   
38158      10033      David     Ochoa  281.867.1615   
49037      10033      David     Ochoa  281.867.1615   
41773      10033      David     Ochoa  281.867.1615   

                                      Email  Order Number  Order Date  \
317    david.measure.3427@morales-young.org       8159112  2023-01-16   
20791  david.measure.3427@morales-young.org       8159112  2023-01-16   
38158  david.measure.3427@morales-young.org       8159112  2023-01-16   
49037  david.measure.3427@morales-young.org       8159112  2023-01-16   
41773  david.measure.3427@morales-young.org       3574869  2023-01-26   

       Order Week  Order Year     Category  ...   Price    Cost  Quantity  \
317             3        2023    furniture  ...  196.08  145.92      1100   
20791           3        2023  consumables  ...   91.05   77.83       327   
38158           3        2023    furniture  ...    1.48    1.22       146   
49037           3        2023        decor  ...  112.34   61.72        10   
41773           4        2023     software  ...  311.98  178.12         6   

       Weight  Line Number   Subtotal  Shipping Price   Total Price  \
317      2.76            1  215688.00            27.6  215690.55300   
20791    0.83            0   29773.35             8.3   29774.11775   
38158   18.04            3     216.08           180.4     232.76700   
49037    8.50            2    1123.40            85.0    1131.26250   
41773    2.34            3    1871.88            23.4    1874.04450   

       Line Cost  Item Cost  
317    160539.60  160539.60  
20791   25458.71   25458.71  
38158     358.52     358.52  
49037     702.20     702.20  
41773    1092.12    1092.12  

[5 rows x 22 columns]

In [ ]:
# Create a column for the profit of each line using line cost and line price
cust_corders_df['Item Profit'] = 


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  shipping_price  line_price  line_cost  \
0      115164.00        787.50         5512.50   131839.08   85597.05   
1         523.95         31.29          312.90      914.26     629.79   
2         527.28         65.52          458.64     1077.12     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [15]:
# Check your work using the totals above


Order 2742071 total: $152811.89
Order 2173913 total: $162388.71
Order 6128929 total: $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.



33615: $8377308.52
66037: $10259514.79
46820: $9743794.36
38378: $12906550.87
24741: $82268892.02


In [17]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 


client_id     qty  shipping_price   line_price    line_cost  line_profit
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31

In [18]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 




Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      33615   64313             1.828985                  8.377309   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
3      38378   73667             3.429455                 12.906551   
4      24741  239862             5.126448                 82.268892   

   Total Cost (millions)  Total Profit (millions)  
0               6.175314                 2.201995  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
3               9.634721                 3.271830  
4              45.688900                36.579992

In [19]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.


Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
4      24741  239862             5.126448                 82.268892   
3      38378   73667             3.429455                 12.906551   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
0      33615   64313             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995